src/stdlib_stats_median.fypp

```diff
@@ -0,0 +1,217 @@
+#:include "common.fypp"
+#:set RANKS = range(1, MAXRANK + 1)
+#:set IR_KINDS_TYPES_OUTPUT = list(zip(REAL_KINDS, REAL_TYPES, REAL_KINDS)) + list(zip(INT_KINDS,INT_TYPES, ['dp']*len(INT_KINDS)))
+
+
+submodule (stdlib_stats) stdlib_stats_median
+
+  use, intrinsic:: ieee_arithmetic, only: ieee_value, ieee_quiet_nan
+  use stdlib_error, only: error_stop
+  use stdlib_optval, only: optval
+  use stdlib_sorting, only:sort => ord_sort
+  implicit none
+
+contains
+
+  #:for k1, t1, o1 in IR_KINDS_TYPES_OUTPUT
+    #:for rank in RANKS
+      #:set RName = rname("median_all",rank, t1, k1, o1)
+      module function ${RName}$ (x, mask) result(res)
+        ${t1}$, intent(in) :: x${ranksuffix(rank)}$
+        logical, intent(in), optional :: mask
+        real(${o1}$) :: res
+
+        integer(kind = int64) :: c, n
+        ${t1}$, allocatable :: x_tmp(:)
+
+        if (.not.optval(mask, .true.) .or. size(x) == 0) then
+          res = ieee_value(1._${o1}$, ieee_quiet_nan)
+          return
+        end if
+
+        n = size(x, kind=int64)
+        c = floor( (n + 1) / 2._${o1}$, kind=int64 )
+
+        x_tmp = reshape(x, [n])
+
+        call sort(x_tmp)
+
+        if (mod(n, 2_int64) == 0) then
+          #:if t1[0] == 'r'
+            res = sum(x_tmp(c:c+1)) / 2._${o1}$
+          #:else
+            res = sum( real(x_tmp(c:c+1), kind=${o1}$) ) / 2._${o1}$
+          #:endif
+        else
+            res = x_tmp(c)
+        end if
+
+      end function ${RName}$
+    #:endfor
+  #:endfor
+
+  #:for k1, t1, o1 in IR_KINDS_TYPES_OUTPUT
+    #:for rank in RANKS
+      #:set RName = rname("median",rank, t1, k1, o1)
+      module function ${RName}$(x, dim, mask) result(res)
+        ${t1}$, intent(in) :: x${ranksuffix(rank)}$
+        integer, intent(in) :: dim
+        logical, intent(in), optional :: mask
+        real(${o1}$) :: res${reduced_shape('x', rank, 'dim')}$
+
+        integer :: c, n
+        #:if rank > 1
+        #:for fj in range(1, rank+1)
+        integer :: j${"_" * fj}$
+        #:endfor
+        #:endif
+        ${t1}$, allocatable :: x_tmp(:)
+
+        if (.not.optval(mask, .true.) .or. size(x) == 0) then
+          res = ieee_value(1._${o1}$, ieee_quiet_nan)
+          return
+        end if
+
+        n = size(x, dim)
+        c = floor( (n + 1) / 2._${o1}$ )
+
+        allocate(x_tmp(n))
+
+        select case(dim)
+          #:for fi in range(1, rank+1)
+          case(${fi}$)
```

Comment:
Here I suggest to add a comment for clarity
```
! Loop over every dimension of the array except "dim"
```

Also, to make it simpler, can you reduce the 2 levels of nested loop construction into 1, with something like the following (untested)?:
```
#: for fj in [range(1, fi), range(fi+1, rank+1)]
```

Reply:
The above should be a 'review' comment, blame my poor github skills...

Reply:
Thank you @gareth-nx for your review. Both done.